File: ports/go/internal/runner/runner.go
package runner

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"reflect"
	"runtime"
	"runtime/debug"
	"sort"
	"strconv"
	"strings"
	"time"
	"unicode/utf16"
)

const checksumModulus = 1_000_000_007

func Run(arguments Arguments) (*Result, error) {
	config := MeasureConfig{}
	if err := readJSON(arguments.Config, &config); err != nil {
		return nil, err
	}
	profile, exists := config.Profiles[arguments.Profile]
	if !exists {
		return nil, fmt.Errorf("unknown profile: %s", arguments.Profile)
	}

	port, err := findPort(config.Ports, "go")
	if err != nil {
		return nil, err
	}
	actualParserVersion, err := dependencyVersion(port.Parser.Module)
	if err != nil {
		return nil, err
	}
	if actualParserVersion != port.Parser.Version {
		return nil, fmt.Errorf("configured %s %s, loaded %s", port.Parser.Module, port.Parser.Version, actualParserVersion)
	}

	manifests, err := loadManifests(arguments.Benchmarks)
	if err != nil {
		return nil, err
	}
	capabilities := make([]CapabilityGroup, 0, len(manifests))
	measurements := make([]Measurement, 0)
	for _, manifest := range manifests {
		parser, parserErr := makeParser(manifest.ID)
		if parserErr != nil {
			return nil, parserErr
		}
		capabilities = append(capabilities, runCapabilities(manifest, parser))
		for _, performanceCase := range manifest.PerformanceCases {
			measurement, measurementErr := runMeasurement(manifest, performanceCase, parser, profile)
			if measurementErr != nil {
				return nil, measurementErr
			}
			measurements = append(measurements, measurement)
		}
	}

	environment := detectEnvironment(arguments.HostFingerprint)
	return &Result{
		Schema:        "https://tabnas.github.io/measure/schemas/port-result.schema.json",
		SchemaVersion: 3,
		Run: RunMetadata{
			ID: arguments.RunID, GeneratedAt: arguments.GeneratedAt,
			Profile: arguments.Profile, SuiteVersion: config.SuiteVersion,
			RepositoryCommit: arguments.Commit, RepositoryDirty: arguments.Dirty,
		},
		Port: PortMetadata{
			ID: port.ID, Label: port.Label, Language: "Go", Runtime: "Go",
			RuntimeVersion: runtime.Version(), ParserModule: port.Parser.Module,
			ParserVersion: actualParserVersion,
		},
		Environment: environment,
		Methodology: Methodology{
			Scope: "parse-only-steady-state-sequential", WarmupMS: profile.WarmupMS,
			SampleTargetMS: profile.SampleTargetMS, Samples: profile.Samples,
			MaxIterations: profile.MaxIterations,
		},
		Capabilities: capabilities,
		Measurements: measurements,
	}, nil
}

// RunDeterministic mirrors the Rust runner's mode of the same name: no
// warmup, no calibration, no clock, one parse and then exactly Iterations
// parses of one case, with both checksums in the output. The harness
// runs the process under callgrind at the configured count and again at
// zero, and the difference is what the counted parses cost; the config
// read, the parser construction, the input generation and the parse
// before the loop are in both and cancel. Valgrind only tolerates this
// binary with GOMAXPROCS=1, which the harness sets, and the count is
// only repeatable with GOGC=off, which it also sets; the document
// reports both as the runtime has them, not as the environment says.
func RunDeterministic(arguments DeterministicArguments) (*DeterministicResult, error) {
	config := MeasureConfig{}
	if err := readJSON(arguments.Config, &config); err != nil {
		return nil, err
	}
	port, err := findPort(config.Ports, "go")
	if err != nil {
		return nil, err
	}
	actualParserVersion, err := dependencyVersion(port.Parser.Module)
	if err != nil {
		return nil, err
	}
	if actualParserVersion != port.Parser.Version {
		return nil, fmt.Errorf("configured %s %s, loaded %s", port.Parser.Module, port.Parser.Version, actualParserVersion)
	}

	manifests, err := loadManifests(arguments.Benchmarks)
	if err != nil {
		return nil, err
	}
	var manifest *BenchmarkManifest
	for index := range manifests {
		if manifests[index].ID == arguments.BenchmarkID {
			manifest = &manifests[index]
		}
	}
	if manifest == nil {
		return nil, fmt.Errorf("unknown benchmark: %s", arguments.BenchmarkID)
	}
	var performanceCase *PerformanceCase
	for index := range manifest.PerformanceCases {
		if manifest.PerformanceCases[index].ID == arguments.CaseID {
			performanceCase = &manifest.PerformanceCases[index]
		}
	}
	if performanceCase == nil {
		return nil, fmt.Errorf("unknown performance case: %s/%s", arguments.BenchmarkID, arguments.CaseID)
	}
	return countParses(manifest, performanceCase, arguments.Iterations)
}

// countParses is the counted mode proper, from the manifest and case
// already resolved: the parser, the input, one parse, the loop, and the
// document. It is what the package's tests exercise, because the parser
// version check above reads build information a test binary does not
// carry; the built binary's whole path is covered by
// scripts/runners.test.mjs.
func countParses(manifest *BenchmarkManifest, performanceCase *PerformanceCase, iterations int) (*DeterministicResult, error) {
	parser, err := makeParser(manifest.ID)
	if err != nil {
		return nil, err
	}
	input, err := generateInput(*performanceCase)
	if err != nil {
		return nil, err
	}
	environment := effectiveSettings()

	// One parse before the counted loop, in both runs. Whatever an engine
	// leaves until it is first asked to parse is then in the baseline as
	// well as in the measured run, and cancels, instead of sitting in the
	// per-parse figure at one part in Iterations. Its checksum is reported
	// on its own so the harness can hold the loop's to it.
	first, parseErr := parser.Parse(input)
	if parseErr != nil {
		return nil, fmt.Errorf("%s/%s parse: %w", manifest.ID, performanceCase.ID, parseErr)
	}
	parseChecksum := int64(checksumValue(first))

	checksum := int64(0)
	for index := 0; index < iterations; index++ {
		result, parseErr := parser.Parse(input)
		if parseErr != nil {
			return nil, fmt.Errorf("%s/%s parse: %w", manifest.ID, performanceCase.ID, parseErr)
		}
		checksum = (checksum + int64(checksumValue(result))) % checksumModulus
	}

	return &DeterministicResult{
		BenchmarkID: manifest.ID, CaseID: performanceCase.ID,
		Input: InputIdentity{
			Bytes: len([]byte(input)), CodeUnits: len(utf16.Encode([]rune(input))), SHA256: hashString(input),
		},
		Iterations: iterations, ParseChecksum: float64(parseChecksum), Checksum: float64(checksum),
		Environment: environment,
	}, nil
}

// The settings the count depends on, read back from the runtime rather
// than copied from the environment: GOMAXPROCS as the scheduler has it
// and GOGC as the collector has it. The document then says what the
// process ran under, and the harness holds that against what the config
// asked for, so a count taken with the collector on cannot be recorded
// under a config that says it was off.
//
// SetGCPercent is the runtime's only reader of the percent that can say
// "off": runtime/metrics reports the same state as a wrapped uint64. It
// returns the previous value, and setting that straight back leaves the
// runtime as it was. Called before any parse, so both counted runs carry
// it alike.
func effectiveSettings() map[string]string {
	percent := debug.SetGCPercent(-1)
	debug.SetGCPercent(percent)
	gogc := strconv.Itoa(percent)
	if percent < 0 {
		gogc = "off"
	}
	return map[string]string{
		"GOMAXPROCS": strconv.Itoa(runtime.GOMAXPROCS(0)),
		"GOGC":       gogc,
	}
}

func runCapabilities(manifest BenchmarkManifest, parser parserAdapter) CapabilityGroup {
	results := make([]CapabilityResult, 0, len(manifest.CapabilityCases))
	passed := 0
	for _, testCase := range manifest.CapabilityCases {
		actual, parseErr := parser.Parse(testCase.Input)
		accepted := parseErr == nil
		casePassed := accepted == testCase.Accept && (!testCase.Accept || reflect.DeepEqual(actual, testCase.Expected))
		result := CapabilityResult{
			CaseID: testCase.ID, Description: testCase.Description,
			AcceptExpected: testCase.Accept, Accepted: accepted, Passed: casePassed,
		}
		if testCase.Accept {
			result.Expected = testCase.Expected
		}
		if accepted {
			result.Actual = actual
		}
		if parseErr != nil {
			result.Error = cleanError(parseErr)
		}
		if casePassed {
			passed++
		}
		results = append(results, result)
	}
	return CapabilityGroup{BenchmarkID: manifest.ID, Passed: passed, Total: len(results), Cases: results}
}

func runMeasurement(manifest BenchmarkManifest, performanceCase PerformanceCase, parser parserAdapter, profile Profile) (Measurement, error) {
	input, err := generateInput(performanceCase)
	if err != nil {
		return Measurement{}, err
	}
	if _, err = parser.Parse(input); err != nil {
		return Measurement{}, fmt.Errorf("%s/%s correctness parse: %w", manifest.ID, performanceCase.ID, err)
	}

	warmupUntil := time.Now().Add(time.Duration(profile.WarmupMS) * time.Millisecond)
	for time.Now().Before(warmupUntil) {
		if _, err = parser.Parse(input); err != nil {
			return Measurement{}, err
		}
	}

	iterations := 1
	for iterations < profile.MaxIterations {
		elapsed, _, batchErr := timeBatch(parser, input, iterations)
		if batchErr != nil {
			return Measurement{}, batchErr
		}
		if elapsed >= time.Duration(profile.SampleTargetMS)*time.Millisecond {
			break
		}
		iterations *= 2
		if iterations > profile.MaxIterations {
			iterations = profile.MaxIterations
		}
	}

	runtime.GC()
	samples := make([]Sample, 0, profile.Samples)
	checksum := 0.0
	for index := 0; index < profile.Samples; index++ {
		elapsed, batchChecksum, batchErr := timeBatch(parser, input, iterations)
		if batchErr != nil {
			return Measurement{}, batchErr
		}
		checksum = float64((int64(checksum) + int64(batchChecksum)) % checksumModulus)
		samples = append(samples, Sample{ElapsedNS: float64(elapsed.Nanoseconds())})
	}

	return Measurement{
		BenchmarkID: manifest.ID, CaseID: performanceCase.ID,
		Description: performanceCase.Description,
		Input: InputIdentity{
			Bytes: len([]byte(input)), CodeUnits: len(utf16.Encode([]rune(input))), SHA256: hashString(input),
		},
		IterationsPerSample: iterations, Samples: samples, Checksum: checksum,
	}, nil
}

func timeBatch(parser parserAdapter, input string, iterations int) (time.Duration, float64, error) {
	checksum := int64(0)
	started := time.Now()
	for index := 0; index < iterations; index++ {
		result, err := parser.Parse(input)
		if err != nil {
			return 0, 0, err
		}
		checksum = (checksum + int64(checksumValue(result))) % checksumModulus
	}
	return time.Since(started), float64(checksum), nil
}

func checksumValue(value any) int {
	switch typed := value.(type) {
	case float64:
		return int(typed) % checksumModulus
	case int:
		return typed % checksumModulus
	case bool:
		if typed {
			return 1
		}
		return 0
	default:
		encoded, _ := json.Marshal(value)
		return len(encoded)
	}
}

func generateInput(performanceCase PerformanceCase) (string, error) {
	generator := performanceCase.Generator
	switch generator.Kind {
	case "adder-chain":
		terms := make([]string, generator.Size)
		for index := range terms {
			terms[index] = strconv.Itoa(generator.Value)
		}
		return strings.Join(terms, "+"), nil
	case "even-palindrome":
		if generator.Size <= 0 || generator.Size%2 != 0 || generator.Pattern == "" {
			return "", fmt.Errorf("invalid even-palindrome generator for %s", performanceCase.ID)
		}
		halfLength := generator.Size / 2
		half := strings.Repeat(generator.Pattern, (halfLength+len(generator.Pattern)-1)/len(generator.Pattern))[:halfLength]
		halfRunes := []rune(half)
		reversed := make([]rune, len(halfRunes))
		for index := range halfRunes {
			reversed[len(halfRunes)-1-index] = halfRunes[index]
		}
		return half + string(reversed), nil
	default:
		return "", fmt.Errorf("unknown generator kind: %s", generator.Kind)
	}
}

func loadManifests(directory string) ([]BenchmarkManifest, error) {
	paths, err := filepath.Glob(filepath.Join(directory, "*", "benchmark.json"))
	if err != nil {
		return nil, err
	}
	sort.Strings(paths)
	manifests := make([]BenchmarkManifest, 0, len(paths))
	for _, path := range paths {
		manifest := BenchmarkManifest{}
		if err = readJSON(path, &manifest); err != nil {
			return nil, err
		}
		manifests = append(manifests, manifest)
	}
	return manifests, nil
}

func readJSON(path string, destination any) error {
	content, err := os.ReadFile(path)
	if err != nil {
		return fmt.Errorf("read %s: %w", path, err)
	}
	if err = json.Unmarshal(content, destination); err != nil {
		return fmt.Errorf("decode %s: %w", path, err)
	}
	return nil
}

func findPort(ports []PortConfig, id string) (PortConfig, error) {
	for _, port := range ports {
		if port.ID == id {
			return port, nil
		}
	}
	return PortConfig{}, fmt.Errorf("the configuration has no %s port", id)
}

func dependencyVersion(module string) (string, error) {
	info, ok := debug.ReadBuildInfo()
	if !ok {
		return "", errors.New("Go build information is unavailable")
	}
	for _, dependency := range info.Deps {
		if dependency.Path == module {
			return strings.TrimPrefix(dependency.Version, "v"), nil
		}
	}
	return "", fmt.Errorf("dependency %s is absent from Go build information", module)
}

func detectEnvironment(hostFingerprint string) Environment {
	osID, architecture := runtime.GOOS, runtime.GOARCH
	osName := operatingSystemName(osID)
	kernelVersion := operatingSystemKernelVersion()
	cpu := cpuModel()
	logicalCPUs := runtime.NumCPU()
	memoryBytes := totalMemoryBytes()
	fingerprintInput, _ := json.Marshal(struct {
		OS            string `json:"os"`
		OSName        string `json:"osName"`
		KernelVersion string `json:"kernelVersion"`
		Arch          string `json:"arch"`
		CPU           string `json:"cpu"`
		LogicalCPUs   int    `json:"logicalCpus"`
		MemoryBytes   uint64 `json:"memoryBytes"`
	}{osID, osName, kernelVersion, architecture, cpu, logicalCPUs, memoryBytes})
	return Environment{
		Fingerprint: hashString(string(fingerprintInput)), HostFingerprint: hostFingerprint,
		OS: osID, OSName: osName,
		KernelVersion: kernelVersion, Arch: architecture, CPU: cpu,
		LogicalCPUs: logicalCPUs, MemoryBytes: memoryBytes,
	}
}

func operatingSystemName(fallback string) string {
	for _, path := range []string{"/etc/os-release", "/usr/lib/os-release"} {
		content, err := os.ReadFile(path)
		if err != nil {
			continue
		}
		for _, line := range strings.Split(string(content), "\n") {
			if value, found := strings.CutPrefix(line, "PRETTY_NAME="); found {
				if name := unquoteOSReleaseValue(value); name != "" {
					return name
				}
			}
		}
	}
	return fallback
}

func unquoteOSReleaseValue(value string) string {
	value = strings.TrimSpace(value)
	if len(value) >= 2 && ((value[0] == '"' && value[len(value)-1] == '"') ||
		(value[0] == '\'' && value[len(value)-1] == '\'')) {
		value = value[1 : len(value)-1]
	}
	value = strings.ReplaceAll(value, `\"`, `"`)
	return strings.ReplaceAll(value, `\\`, `\`)
}

func operatingSystemKernelVersion() string {
	if content, err := os.ReadFile("/proc/sys/kernel/osrelease"); err == nil {
		if version := strings.TrimSpace(string(content)); version != "" {
			return version
		}
	}
	if output, err := exec.Command("uname", "-r").Output(); err == nil {
		if version := strings.TrimSpace(string(output)); version != "" {
			return version
		}
	}
	return runtime.GOOS
}

func cleanError(err error) string {
	message := strings.Split(err.Error(), "\n")[0]
	if len(message) > 500 {
		return message[:500]
	}
	return message
}

func hashString(value string) string {
	digest := sha256.Sum256([]byte(value))
	return hex.EncodeToString(digest[:])
}

File: ports/rust/src/runner.rs
// The measurement loop. It mirrors `ports/go/internal/runner/runner.go` and
// `ports/typescript/src/main.ts` step for step: same warmup, same iteration
// calibration, same checksum arithmetic, same input identity. Anything that
// differs between the three would show up as a divergent number that looks
// like a language result and is not one.

use std::collections::BTreeMap;
use std::fs;
use std::path::Path;
use std::time::{Duration, Instant};

use serde_json::Value as Json;
use sha2::{Digest, Sha256};
use tabnas::{Tabnas, Value};

use crate::environment;
use crate::model::{
    Arguments, BenchmarkManifest, CapabilityGroup, CapabilityResult, InputIdentity, MeasureConfig,
    Measurement, Methodology, PerformanceCase, PortMetadata, PortResult, Profile, RunMetadata,
    Sample,
};
use crate::parsers::make_parser;

const CHECKSUM_MODULUS: i64 = 1_000_000_007;
const PORT_ID: &str = "rust";
const SCHEMA: &str = "https://tabnas.github.io/measure/schemas/port-result.schema.json";

pub fn run(arguments: &Arguments) -> Result<PortResult, String> {
    let config: MeasureConfig = read_json(&arguments.config)?;
    let profile = *config
        .profiles
        .get(&arguments.profile)
        .ok_or_else(|| format!("unknown profile: {}", arguments.profile))?;

    let port = config
        .ports
        .iter()
        .find(|candidate| candidate.id == PORT_ID)
        .ok_or_else(|| format!("the configuration has no {PORT_ID} port"))?;
    if tabnas::VERSION != port.parser.version {
        return Err(format!(
            "configured {} {}, loaded {}",
            port.parser.module,
            port.parser.version,
            tabnas::VERSION
        ));
    }

    let manifests = load_manifests(&arguments.benchmarks)?;
    let mut capabilities = Vec::with_capacity(manifests.len());
    let mut measurements = Vec::new();
    for manifest in &manifests {
        let parser = make_parser(&manifest.id)?;
        capabilities.push(run_capabilities(manifest, &parser));
        for performance_case in &manifest.performance_cases {
            measurements.push(run_measurement(manifest, performance_case, &parser, &profile)?);
        }
    }

    Ok(PortResult {
        schema: SCHEMA,
        schema_version: 3,
        run: RunMetadata {
            id: arguments.run_id.clone(),
            generated_at: arguments.generated_at.clone(),
            profile: arguments.profile.clone(),
            suite_version: config.suite_version.clone(),
            repository_commit: arguments.commit.clone(),
            repository_dirty: arguments.dirty,
        },
        port: PortMetadata {
            id: port.id.clone(),
            label: port.label.clone(),
            language: "Rust",
            runtime: "Rust",
            runtime_version: env!("MEASURE_RUSTC_VERSION").to_string(),
            parser_module: port.parser.module.clone(),
            parser_version: tabnas::VERSION.to_string(),
        },
        environment: environment::detect(&arguments.host_fingerprint),
        methodology: Methodology {
            scope: "parse-only-steady-state-sequential",
            warmup_ms: profile.warmup_ms,
            sample_target_ms: profile.sample_target_ms,
            samples: profile.samples,
            max_iterations: profile.max_iterations,
        },
        capabilities,
        measurements,
    })
}

fn run_capabilities(manifest: &BenchmarkManifest, parser: &Tabnas) -> CapabilityGroup {
    let mut cases = Vec::with_capacity(manifest.capability_cases.len());
    let mut passed = 0;
    for test_case in &manifest.capability_cases {
        let outcome = parser.parse(&test_case.input);
        let accepted = outcome.is_ok();
        let actual = outcome.as_ref().ok().map(canonical_json);
        // Present-and-null is a value to compare against; absent is not.
        let expected = test_case.expected.as_ref().map(|value| match value {
            Some(value) => normalize(value),
            None => Json::Null,
        });
        // Both sides are normalized, so a whole-valued `2.0` from the
        // parser and a `2` in the manifest agree the way they do in the
        // other two ports.
        let case_passed =
            accepted == test_case.accept && (!test_case.accept || actual == expected);
        if case_passed {
            passed += 1;
        }
        cases.push(CapabilityResult {
            case_id: test_case.id.clone(),
            description: test_case.description.clone(),
            accept_expected: test_case.accept,
            accepted,
            passed: case_passed,
            // The manifest's own value, as the other two runners emit it:
            // present when the manifest has the key, null when it says null.
            expected: test_case
                .expected
                .clone()
                .map(|value| value.unwrap_or(Json::Null)),
            actual,
            error: outcome.err().map(|error| clean_error(&error.to_string())),
        });
    }
    CapabilityGroup {
        benchmark_id: manifest.id.clone(),
        passed,
        total: cases.len(),
        cases,
    }
}

fn run_measurement(
    manifest: &BenchmarkManifest,
    performance_case: &PerformanceCase,
    parser: &Tabnas,
    profile: &Profile,
) -> Result<Measurement, String> {
    let input = generate_input(performance_case)?;
    parser.parse(&input).map_err(|error| {
        format!(
            "{}/{} correctness parse: {}",
            manifest.id,
            performance_case.id,
            clean_error(&error.to_string())
        )
    })?;

    let warmup_until = Instant::now() + Duration::from_millis(profile.warmup_ms);
    while Instant::now() < warmup_until {
        parse_or_fail(parser, &input)?;
    }

    let target = Duration::from_millis(profile.sample_target_ms);
    let mut iterations = 1;
    while iterations < profile.max_iterations {
        let (elapsed, _) = time_batch(parser, &input, iterations)?;
        if elapsed >= target {
            break;
        }
        iterations = (iterations * 2).min(profile.max_iterations);
    }

    // The Go and TypeScript runners collect garbage here. Rust frees the
    // previous parse as it goes, so there is no deferred work to flush and
    // nothing to do at this point.
    let mut samples = Vec::with_capacity(profile.samples);
    let mut checksum = 0.0_f64;
    for _ in 0..profile.samples {
        let (elapsed, batch_checksum) = time_batch(parser, &input, iterations)?;
        checksum = ((checksum as i64 + batch_checksum as i64) % CHECKSUM_MODULUS) as f64;
        samples.push(Sample {
            elapsed_ns: elapsed.as_nanos() as f64,
        });
    }

    Ok(Measurement {
        benchmark_id: manifest.id.clone(),
        case_id: performance_case.id.clone(),
        description: performance_case.description.clone(),
        input: input_identity(&input),
        iterations_per_sample: iterations,
        samples,
        checksum,
    })
}

fn time_batch(parser: &Tabnas, input: &str, iterations: usize) -> Result<(Duration, f64), String> {
    let mut checksum = 0_i64;
    let started = Instant::now();
    for _ in 0..iterations {
        let value = parse_or_fail(parser, input)?;
        checksum = (checksum + checksum_value(&value)) % CHECKSUM_MODULUS;
    }
    Ok((started.elapsed(), checksum as f64))
}

fn parse_or_fail(parser: &Tabnas, input: &str) -> Result<Value, String> {
    parser
        .parse(input)
        .map_err(|error| clean_error(&error.to_string()))
}

fn checksum_value(value: &Value) -> i64 {
    match value {
        Value::Number(number) => (*number as i64) % CHECKSUM_MODULUS,
        Value::Bool(flag) => i64::from(*flag),
        other => serde_json::to_string(&canonical_json(other))
            .map(|encoded| encoded.len() as i64)
            .unwrap_or(0),
    }
}

fn generate_input(performance_case: &PerformanceCase) -> Result<String, String> {
    let generator = &performance_case.generator;
    match generator.kind.as_str() {
        "adder-chain" => Ok(vec![generator.value.to_string(); generator.size].join("+")),
        "even-palindrome" => {
            if generator.size == 0 || generator.size % 2 != 0 || generator.pattern.is_empty() {
                return Err(format!(
                    "invalid even-palindrome generator for {}",
                    performance_case.id
                ));
            }
            let half_length = generator.size / 2;
            let pattern_length = generator.pattern.chars().count();
            let repeats = half_length.div_ceil(pattern_length);
            let half: String = generator
                .pattern
                .repeat(repeats)
                .chars()
                .take(half_length)
                .collect();
            let reversed: String = half.chars().rev().collect();
            Ok(format!("{half}{reversed}"))
        }
        kind => Err(format!("unknown generator kind: {kind}")),
    }
}

fn load_manifests(directory: &str) -> Result<Vec<BenchmarkManifest>, String> {
    let entries = fs::read_dir(directory)
        .map_err(|error| format!("read {directory}: {error}"))?
        .collect::<Result<Vec<_>, _>>()
        .map_err(|error| format!("read {directory}: {error}"))?;
    // Keyed by benchmark id rather than by directory name: the aggregator
    // pairs ports by id, and the TypeScript runner sorts by id too.
    let mut manifests = BTreeMap::new();
    for entry in entries {
        if !entry.path().is_dir() {
            continue;
        }
        let path = entry.path().join("benchmark.json");
        let manifest: BenchmarkManifest = read_json(&path.to_string_lossy())?;
        manifests.insert(manifest.id.clone(), manifest);
    }
    Ok(manifests.into_values().collect())
}

fn read_json<T: serde::de::DeserializeOwned>(path: &str) -> Result<T, String> {
    let content =
        fs::read_to_string(Path::new(path)).map_err(|error| format!("read {path}: {error}"))?;
    serde_json::from_str(&content).map_err(|error| format!("decode {path}: {error}"))
}

pub fn input_identity(input: &str) -> InputIdentity {
    InputIdentity {
        bytes: input.len(),
        code_units: input.encode_utf16().count(),
        sha256: sha256(input),
    }
}

pub fn sha256(value: &str) -> String {
    let digest = Sha256::digest(value.as_bytes());
    digest.iter().map(|byte| format!("{byte:02x}")).collect()
}

/// Convert a parse result to JSON the way `encoding/json` and
/// `JSON.stringify` do. Both write a whole-valued number without a
/// fractional part, and the aggregator compares ports by their serialized
/// text, so `6.0` here would read as a Rust-only divergence.
pub fn canonical_json(value: &Value) -> Json {
    normalize(&value.to_json())
}

fn normalize(value: &Json) -> Json {
    match value {
        Json::Number(number) => match number.as_f64() {
            Some(number) if number.fract() == 0.0 && number.abs() < 9.007_199_254_740_992e15 => {
                Json::Number((number as i64).into())
            }
            _ => value.clone(),
        },
        Json::Array(items) => Json::Array(items.iter().map(normalize).collect()),
        Json::Object(entries) => Json::Object(
            entries
                .iter()
                .map(|(key, entry)| (key.clone(), normalize(entry)))
                .collect(),
        ),
        _ => value.clone(),
    }
}

fn clean_error(message: &str) -> String {
    let mut plain = String::with_capacity(message.len());
    let mut characters = message.chars().peekable();
    // Diagnostics may be coloured. The raw result is compared and stored as
    // text, so the escape sequences come out here rather than in a reader's
    // terminal.
    while let Some(character) = characters.next() {
        if character == '\u{1b}' && characters.peek() == Some(&'[') {
            for tail in characters.by_ref() {
                if tail.is_ascii_alphabetic() {
                    break;
                }
            }
            continue;
        }
        plain.push(character);
    }
    plain
        .lines()
        .next()
        .unwrap_or("parse error")
        .chars()
        .take(500)
        .collect()
}
